glib/tests/param_spec.rs

```diff
@@ -0,0 +1,11 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+#[test]
+fn generated_builder() {
+    let spec = glib::ParamSpecCharBuilder::new()
+        .name("custom-char")
```

Comment:
What happens if you don't provide a name? I believe that one should be passed to the constructor. 

A bit unrelated, but is a proc macro really needed here ?

Reply:
Name, nick, blurb and flags are mandatory so those should really be passed in the constructor.

Also there should be a `ParamSpecFoo::builder()` function that returns the builder.

And I would prefer to not have this as proc macro as that automatically becomes public API that we'll have to maintain.

Reply:
Otherwise this seems like an useful API addition

Reply:
> Name, nick, blurb and flags are mandatory so those should really be passed in the constructor.

It currently assumes nick/blurb are equal to name which is what Vala/PyGObject seems to do iirc

Reply:
That makes sense

Reply:
Something to note about the auto-generated builder pattern is  there is no way to document it in the future which is imho not great

Reply:
Yeah as there's a limited number of pspecs it doesn't seem worth it

Reply:
> What happens if you don't provide a name? 

It panics with `Missing parameter name`. The same treatment goes to every parameter (unless there's a default value set in the macro invocation).

> I would prefer to not have this as proc macro as that automatically becomes public API that we'll have to maintain.

Well, yes, but this is also useful for external users. For example, gstreamer has some `ParamSpec`s that could take advantage of this macro.

> there is no way to document it in the future which is imho not great

That's true, but ideally why should a `ParamSpec*Builder` get specific documentation? If the documentation is common between Specs, it can be generated by the macro (for example, documentation about the default values, which I missed).

> [...] there's a limited number of pspecs it doesn't seem worth it.

Well, I could manually write everything, but then you'll get a commit with an addition of ~2000 lines. This only adds ~300 lines, and makes every `ParamSpec*Builder` behave the same way. Maintaining those 2000 lines manually could be tricky.

Reply:
You don't have to write everything by hand, the ParamSpec* in glib already make use of macros to generate the common parts. You could something similar here I guess.

Reply:
Ok, but I'll wait for a final feedback from @sdroege before rewriting everything

Reply:
I agree with @bilelmoussaoui here